### PR TITLE
feat: implement signData for Ledger hardware wallets

### PR DESCRIPTION
**Pull request by AngelCastilloB**
# Context

1. Message signing is now supported by Ledger devices
2. Glacier Drop requires this functionality to claim

# Proposed Solution

Implement message signing for ledger devices

# Important Changes Introduced

Key agents, signer coordinator and witnesser interfaces were refactored to abstract CIP-08 signing

### CODE REVIEW

File: packages/hardware-ledger/src/LedgerKeyAgent.ts [hunk 1]
```diff
@@ -190,6 +208,111 @@ type OpenTransportForDeviceParams = {
...
+): { signingPath: BIP32Path; addressParams: DeviceOwnedAddress } => {
+  if (Cardano.DRepID.isValid(signWith)) {
+    const path = util.accountKeyDerivationPathToBip32Path(accountIndex, util.DREP_KEY_DERIVATION_PATH, purpose);
+
+    return {
+      addressParams: {
+        params: {
+          spendingPath: path
+        },
+        type: AddressType.ENTERPRISE_KEY
```
Comment (rhyslbw) [inline packages/hardware-ledger/src/LedgerKeyAgent.ts:225]: What's this? The payment key?

File: packages/hardware-ledger/src/LedgerKeyAgent.ts [hunk 2]
```diff
@@ -615,8 +737,66 @@ export class LedgerKeyAgent extends KeyAgentBase {
...
+      };
+
+      const deviceConnection = await LedgerKeyAgent.checkDeviceConnection(
+        this.#communicationType,
+        this.deviceConnection
+      );
+
+      const result = await deviceConnection.signMessage(messageData);
+
+      // Re-create the CIP-008 payload the same way the firmware does it internally, otherwise the signature
```
Comment (rhyslbw) [inline packages/hardware-ledger/src/LedgerKeyAgent.ts:769]: What's going on here? This seems like a lot of work and duplication of responsibility on our side 🤔

File: packages/key-management/test/cip8/cip30signData.test.ts [hunk 3]
```diff
@@ -30,25 +19,23 @@ describe('cip30signData', () => {
...
-      sender,
-      signWith,
-      witnesser
+      signWith
     });
 
     const coseKey = COSEKey.from_bytes(Buffer.from(dataSignature.key, 'hex'));
     const coseSign1 = COSESign1.from_bytes(Buffer.from(dataSignature.signature, 'hex'));
 
+    console.log(dataSignature.key);
```
Comment (rhyslbw) [inline packages/key-management/test/cip8/cip30signData.test.ts:37]: These will need eslint ignores if they're intended to stay. Can they not just be assertions?

File: packages/key-management/src/cip8/cip30signData.ts [hunk 4]
```diff
@@ -39,7 +39,7 @@ export class Cip30DataSignError<InnerError = unknown> extends ComposableError<In
   }
 }
 
-const getAddressBytes = (signWith: Cardano.PaymentAddress | Cardano.RewardAccount | Cardano.DRepID) => {
+export const getAddressBytes = (signWith: Cardano.PaymentAddress | Cardano.RewardAccount | Cardano.DRepID) => {
```
Comment (mkazlauskas) [inline packages/key-management/src/cip8/cip30signData.ts:42]: We should add some happy path test for each newly exported function, as it becomes a feature

**Review by mkazlauskas — CHANGES_REQUESTED** [review]

File: packages/hardware-ledger/src/LedgerKeyAgent.ts (hunk 2 above)
Comment (AngelCastilloB) [reply packages/hardware-ledger/src/LedgerKeyAgent.ts:769]: So the way the firmware works is by building the CIP-08 message internally and signing that, but then it only returns the signature and the key used to sign, CIP-30 signData requires that we return the structure used to compute the signature, so our only option is to build it the same way the ledger does it internally as it doesn't return it

File: packages/key-management/test/cip8/cip30signData.test.ts (hunk 3 above)
Comment (AngelCastilloB) [reply packages/key-management/test/cip8/cip30signData.test.ts:37]: ups, I added this for debugging purposes and forgot to remove them

File: packages/hardware-ledger/src/LedgerKeyAgent.ts [hunk 5]
```diff
@@ -190,6 +208,111 @@ type OpenTransportForDeviceParams = {
   device: LedgerDevice;
 };
 
+const getDerivationPath = (
```
[the remark this answers is not in the document]
Comment (AngelCastilloB) [reply packages/hardware-ledger/src/LedgerKeyAgent.ts:211]: We currently dont have something like this in the mappers

File: packages/hardware-ledger/src/LedgerKeyAgent.ts [hunk 6]
```diff
@@ -1,32 +1,50 @@
...
+  cip8,
   errors,
   util
 } from '@cardano-sdk/key-management';
+import { Cip30DataSignature } from '@cardano-sdk/dapp-connector';
 import { HID } from 'node-hid';
+import { HexBlob, areNumbersEqualInConstantTime, areStringsEqualInConstantTime } from '@cardano-sdk/util';
 import { LedgerDevice, LedgerTransportType } from './types';
-import { areNumbersEqualInConstantTime, areStringsEqualInConstantTime } from '@cardano-sdk/util';
+import { MessageAddressFieldType } from '@cardano-foundation/ledgerjs-hw-app-cardano/dist/types/public';
```
[the remark this answers is not in the document]
Comment (AngelCastilloB) [reply packages/hardware-ledger/src/LedgerKeyAgent.ts:34]: It is indexed, I will fix it (probably got auto imported like this by my IDE)

[the remark this answers is not in the document]
Comment (AngelCastilloB) [reply packages/hardware-ledger/src/LedgerKeyAgent.ts]: done here https://github.com/input-output-hk/cardano-js-sdk/pull/1393/commits/fe2a0a833584b471a1341656dd647b7fbafcfa68

File: packages/key-management/test/cip8/cip30signData.test.ts (hunk 3 above)
Comment (AngelCastilloB) [reply packages/key-management/test/cip8/cip30signData.test.ts]: done here https://github.com/input-output-hk/cardano-js-sdk/pull/1393/commits/c132ea363e1ddf7d9fd94d0ce51aeae849d90200

File: packages/hardware-ledger/src/LedgerKeyAgent.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (mkazlauskas) [reply packages/hardware-ledger/src/LedgerKeyAgent.ts]: How about https://github.com/input-output-hk/cardano-js-sdk/pull/1393/files#diff-5f12d0fee65873887295d29801f8fd2cc5b37b68565182065903ea44c04a7511R21 `isRewardAccount`?

File: packages/key-management/src/cip8/cip30signData.ts (hunk 4 above)
Comment (mkazlauskas) [reply packages/key-management/src/cip8/cip30signData.ts:42]: How about https://github.com/input-output-hk/cardano-js-sdk/pull/1393/files#diff-5f12d0fee65873887295d29801f8fd2cc5b37b68565182065903ea44c04a7511R21 `isRewardAccount`?

Comment (AngelCastilloB) [reply packages/key-management/src/cip8/cip30signData.ts:42]: A yes, missed that one, done here:

https://github.com/input-output-hk/cardano-js-sdk/pull/1393/commits/ec2231e59a48f40b1f069d30648d77e90ee0efcd

**Review by mkazlauskas — APPROVED** [review]
Great work! :rocket:

File: packages/hardware-ledger/src/LedgerKeyAgent.ts (hunk 2 above)
Comment (rhyslbw) [reply packages/hardware-ledger/src/LedgerKeyAgent.ts:769]: Urgh, seems like an oversight in the library.

**Review by rhyslbw — APPROVED** [review]
Great work @AngelCastilloB 🧙

**Pull request merged**